iOS 18 Support

jkmassel opened the pull request:
Adds iOS 18 support – we're not building against that SDK yet, but right now this PR:

1. Adds iOS 18 tint-able icons
2. Builds the app with Xcode 16

To test:
Ensure CI passes

jkmassel committed:
Use iOS 18 icon

jkmassel committed:
Build with Xcode 16

dangermattic commented:
<!--
  0 Errors
  3 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/296">25.4</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jkmassel committed:
Fix Swift 6 language mode warnings

jkmassel committed:
Fix ImageDownloader Swift 6 warnings

`cache` is actor-isolated, so it can’t be acessed from a `nonisolated` function without using `await`

I just re-isolated it and updated the callers

jkmassel committed:
Adds a @Sendable conformance so that tests build

jkmassel committed:
Fix an Swift 6 warning

jkmassel committed:
Specify the iOS deployment target

jkmassel committed:
Add retroactive conformance annotations

jkmassel committed:
Added some straightforward sendable conformance

kean committed:
Merge branch 'trunk' into ios-18-changes

kean committed:
Update .xcode-version

kean committed:
Add nonisolated to ImageDownloader caching methods

kean committed:
Address warnings

kean committed:
Silence warnings

kean committed:
Addess more concurrency warnings

kean committed:
Fix more warnings

kean committed:
Fix shared-pipeline-vars

kean committed:
Fix unit tests

kean committed:
Fix warnings in tests

kean committed:
Remove weird unit tests in  WordPressAuthenticator

kean reviewed (approved):
I made some more minor changes and updated the .xcode-version. The icons looks good, and everything else too.

kean committed:
Update unit test device

crazytonyli commented on WordPress/Classes/Utility/ImmuTable.swift:
```diff
@@ -486,6 +486,10 @@ extension UITableViewController: TableViewContainer {}
 // MARK: - Diffable
 
+// This conformance was added during the Xcode 16 migration to silence the
+// dozens of false-positive warnigns (any @unchecked conformance is tech debt).
```
```suggestion
// dozens of false-positive warnings (any @unchecked conformance is tech debt).
```

crazytonyli commented on WordPress/Classes/Utility/ImmuTable.swift:
```diff
@@ -486,6 +486,10 @@ extension UITableViewController: TableViewContainer {}
 
+// This conformance was added during the Xcode 16 migration to silence the
+// dozens of false-positive warnigns (any @unchecked conformance is tech debt).
+extension AnyHashable: @retroactive @unchecked Sendable {}
```
This line has a much bigger impact than the individual `@checked Sendable`: all `AnyHashable` instances now get `Sendable` during runtime, which feels like an unsafe change...

I don't have any solution to solve the warning without changing how the two types below are used. Maybe we can tolerate the warnings for now until a proper `ImmuTable` refactor?

kean replied:
> all AnyHashable instances now get Sendable during runtime, which feels like an unsafe change

It brings the app back to the status quo where in Xcode 15.4 where all `AnyHashable` instances were also not producing any concurrency warnings.

> Maybe we can tolerate the warnings for now until a proper ImmuTable refactor?

Unfortunately no because the vast majority of warnings are due to this type.

<img width="854" alt="e0c0ac73f66f4dd5" src="https://github.com/user-attachments/assets/aecacf60-5434-4120-9998-3b44c789900d">


> I don't have any solution to solve the warning without changing how the two types below are used.

The problem is two-fold:

- `UITableViewDiffableDataSource` requiring `Sendable` even when you never use it from the background (you [can](https://developer.apple.com/documentation/uikit/uitableviewdiffabledatasource/3375811-apply) but very few people actually do). I think it's a pretty major limitation of the design of data race safety features in Swift 6.
- `AnyHashable` can't become `Sendable` conditionally based on the type is is parameterized with because it erases this type at compile time.

The only fix seems to be to stop using `AnyHashable`.

kean committed:
Update WordPress/Classes/Utility/ImmuTable.swift

Co-authored-by: Tony Li <tony.li@automattic.com>

kean committed:
Remove broken tests (no following best practices for unit testing)

kean committed:
Update xcode version

kean committed:
Update names of test devices

kean committed:
Add a workaround for testResetApplicationBadgeWhenAccountChange test

kean committed:
Merge branch 'trunk' into ios-18-changes

kean committed:
Update iPad to run on iOS 17.5 until we can update the tests for the sidebar

kean commented on .buildkite/pipeline.yml:
```diff
@@ -79,7 +79,7 @@ steps:
 
       - label: "🔬 :jetpack: UI Tests (iPad)"
-        command: .buildkite/commands/run-ui-tests.sh 'iPad Pro 13-inch (M4)'
+        command: .buildkite/commands/run-ui-tests.sh 'iPad (10th generation) (17.5)'
```
I stopped trying to get it to work because I have another branch where I remove the .sidebar FF and with a fully updated set of tests.

kean commented on WordPress/WordPressTest/Comments/Controllers/FullScreenCommentReplyViewControllerTests.swift:
```diff
@@ -60,38 +60,6 @@ class FullScreenCommentReplyViewControllerTests: CoreDataTestCase {
-        let view = controller.navigationController?.view.subviews.first { view in
-            return view is SuggestionsTableView
-        }
-        let suggestionsTableView = try XCTUnwrap(view)
```
I spent a good hour trying to debug this but because these tests don't follow best practices for unit testing and are evaluating the UIKit components, I made a call to remove them.

kean commented on WordPress/WordPressTest/Notification/Controllers/NotificationsViewControllerTests.swift:
```diff
@@ -52,6 +52,13 @@ final class NotificationsViewControllerTests: XCTestCase {
         // Then
+        // TODO: rework this unit test
+        let expectation = self.expectation(description: "setBadgeCount")
+        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
```
These tests are failing because the deprecated `applicationIconBadgeNumber` property now works asynchronously. It's not great, but we can rework it when we remove the deprecated API.

kean committed:
Updates tests in testTrafficYearsStatsLoadProperly (for new number formatting)

kean committed:
Temporarily disable tests on iPad
